system-contracts/contracts/PubdataChunkPublisher.sol

```diff
@@ -0,0 +1,38 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.20;
+
+import {IPubdataChunkPublisher} from "./interfaces/IPubdataChunkPublisher.sol";
+import {ISystemContract} from "./interfaces/ISystemContract.sol";
+import {L1_MESSENGER_CONTRACT, BLOB_SIZE_BYTES} from "./Constants.sol";
+import {EfficientCall} from "./libraries/EfficientCall.sol";
+import {SystemContractHelper} from "./libraries/SystemContractHelper.sol";
+import {SystemLogKey} from "./Constants.sol";
+
+/**
+ * @author Matter Labs
+ * @custom:security-contact security@matterlabs.dev
+ * @notice Smart contract for chunking pubdata into the appropriate size for 4844 blobs.
+ */
+contract PubdataChunkPublisher is IPubdataChunkPublisher, ISystemContract {
+    /// @notice Chunks pubdata into pieces that can fit into blobs.
+    /// @param _pubdata The total l2 to l1 pubdata that will be sent via L1 blobs.
+    /// @dev Note: This is an early implementation, in the future we plan to support up to 16 blobs per l1 batch.
+    /// @dev We always publish 2 system logs even if our pubdata fits into a single blob. This makes processing logs on L1 easier.
+    function chunkAndPublishPubdata(bytes calldata _pubdata) external onlyCallFrom(address(L1_MESSENGER_CONTRACT)) {
+        require(_pubdata.length <= BLOB_SIZE_BYTES * 2, "pubdata should fit in 2 blobs");
+        // TODO: Update for dynamic number of hashes:
+        //       blobHashes = new bytes32[]((_pubdata.length / BLOB_SIZE_BYTES) + 1);
+        bytes32[] memory blobHashes = new bytes32[](2);
```

Comment:
Depending on the complexity of the implementation, it would be very good if we could gracefully handle <= 2 number of blobs.

Note, the crypto side requires that we use two blobs, 31 * 4096 bytes each. The option of "zero" is available only if the EIP4844 is turned off completely. https://github.com/matter-labs/era-zkevm_circuits/blob/5076a9a8cd775c8f7a84507a02af1e2350e3679d/src/scheduler/mod.rs#L1045.

One of the approaches I am thinking about is the following one:

- In case, the second blob appears to be empty, we just send a constant empty blob hash to L1. (this constant is `keccak([0u8; 4096 * 31]`.
- On L1, if we see such a hash, we require that the corresponding versioned blob hash is 0 (and don't use the evaluation precompile). 

This way, an external observer can just use all the blobs they see & we don't need to spend additional money on sending the second blob + the implementation sounds easy on the contracts' side

Reply:
does "turned off" mean we decided to send pubdata via calldata?

Reply:
we can use the blob commitment instead of not calling the point eval precompile since if we do it means that we received enough commitment data to try and make the 2nd point evaluation call
```
require(
            (_blobHashes[1] == EMPTY_BLOB_HASH && blobCommitments[1] == bytes32(0)) ||
            (_blobHashes[1] != EMPTY_BLOB_HASH && blobCommitments[1] != bytes32(0)),
            "bh"
        );
```

Reply:
I am not sure I understand your proposal. Unless I miss something in the current code `blobCommitments` can never be zero

https://github.com/matter-labs/era-contracts/pull/138#discussion_r1451072657

Reply:
The blob is empty then we won’t have commitments for it as part of the pubdata and we wont perform the blob commitment calculation

Reply:
when we have 1 blob: the `pubdataCommitments` will be `1 header bytes || opening point (16 bytes) || claimed value (32 bytes) || commitment (48 bytes) || proof (48 bytes))` so the loop will only run once

Reply:
if we did have extra data in there it would fail on the point evaluation precompile as the first step is verifying `versioned_hash == hash(commitment)` which would fail since the `versioned_hash <- bytes32(0)`